keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1504 @@
+    // back to pending.
+    ResourceClaimReallocate ResourceClaimPhase = “Reallocate”
+
+    // Deleting the ResourceClaim triggers freeing the resource.
+    // Because of the plugin’s finalizer, such a claim then
+    // continues to exist with a DeletionTimeStamp.
+    // Once the plugin has successfully freed the resource,
+    // it sets this status and removes the finalizer. Usually
+    // the claim object will then be removed quickly, but
+    // additional finalizers might also keep it around longer.
+    ResourceClaimFreed ResourceClaimPhase = “Freed”
+)
+
+type PodSpec {
+   ...
+   Resources []PodResource
+   ...
+}
+
+type Container {
+   ...
+   // The entries are the names of resources in PodSpec.Resources
+   // that are used by the container.
+   PodResources []string
+   ...
+}
+
+type PodResource struct {
+   // A name under which this resource can be referenced by the containers.
+   Name string
+
+   // The resource is independent of the Pod and defined by
+   // a separate ResourceClaim in the same namespace as
+   // the Pod. Either this or Template must be set, but not both.
+   ResourceClaimName *string
+
+    // Will be used to create a stand-alone ResourceClaim to allocate the resource.
+    // The pod in which this PodResource is embedded will be the
+    // owner of the ResourceClaim, i.e. the ResourceClaim will be deleted together with the
+    // pod.  The name of the ResourceClaim will be `<pod name>-<resource name>` where
+    // `<resource name>` is the name PodResource.Name
+    // Pod validation will reject the pod if the concatenated name
+    // is not valid for a ResourceClaim (for example, too long).
+    //
+    // An existing ResourceClaim with that name that is not owned by the pod
+    // will *not* be used for the pod to avoid using an unrelated
+    // resource by mistake. Starting the pod is then blocked until
+    // the unrelated ResourceClaim is removed. If such a pre-created ResourceClaim is
+    // meant to be used by the pod, the ResourceClaim has to be updated with an
+    // owner reference to the pod once the pod exists. Normally
+    // this should not be necessary, but it may be useful when
+    // manually reconstructing a broken cluster.
+    //
+    // This field is read-only and no changes will be made by Kubernetes
+    // to the ResourceClaim after it has been created.
+   Template *ResourceClaimTemplate
+}
+
+type ResourceClaimTemplate struct {
+    // May contain labels and annotations that will be copied into the ResourceClaim
+    // when creating it. No other fields are allowed and will be rejected during
+    // validation.
+    metav1.ObjectMeta
+
+    // The specification for the ResourceClaim. The entire content is
+    // copied unchanged into the ResourceClaim that gets created from this
+    // template. The same fields as in a ResourceClaim
+    // are also valid here.
+    Spec ResourceClaimSpec
+}
+```
+
+### Communication between kubelet and resource node plugin
```

Comment:
A suggestion: would be nice to have a diagram that clearly shows the endpoint and the full view of gRPC communication between kubelet and the plugin.

Reply:
I agree, this low-level part needs more work. It's not fully spec-ed out.

Reply:
I just pushed a major update with a more complete implementation section.

@bart0sh is going to add more details to the kubelet section.